File: app/src/main/java/org/monksanctum/xand11/fonts/FontProtocol.java
// Copyright 2018 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package org.monksanctum.xand11.fonts;

import android.graphics.Rect;
import android.util.Log;

import org.monksanctum.xand11.Client;
import org.monksanctum.xand11.Dispatcher;
import org.monksanctum.xand11.comm.Event;
import org.monksanctum.xand11.comm.PacketReader;
import org.monksanctum.xand11.comm.PacketWriter;
import org.monksanctum.xand11.comm.Request;
import org.monksanctum.xand11.errors.GContextError;
import org.monksanctum.xand11.errors.XError;
import org.monksanctum.xand11.fonts.Font.CharInfo;
import org.monksanctum.xand11.fonts.Font.FontProperty;
import org.monksanctum.xand11.graphics.GraphicsManager;

import java.util.List;

import static org.monksanctum.xand11.fonts.FontManager.DEBUG;

public class FontProtocol implements Dispatcher.PacketHandler {

    private static final byte[] HANDLED_OPS = new byte[] {
            Request.OPEN_FONT,
            Request.QUERY_FONT,
            Request.CLOSE_FONT,
            Request.LIST_FONTS,
            Request.LIST_FONTS_WITH_INFO,
            Request.QUERY_TEXT_EXTENTS,
    };

    private final FontManager mFontManager;
    private final GraphicsManager mGraphicsManager;

    public FontProtocol(FontManager manager, GraphicsManager graphicsManager) {
        mFontManager = manager;
        mGraphicsManager = graphicsManager;
    }

    @Override
    public byte[] getOpCodes() {
        return HANDLED_OPS;
    }

    @Override
    public void handleRequest(Client client, PacketReader reader, PacketWriter writer)
            throws XError {
        switch (reader.getMajorOpCode()) {
            case Request.OPEN_FONT:
                handleOpenFont(reader);
                break;
            case Request.CLOSE_FONT:
                handleCloseFont(reader);
                break;
            case Request.QUERY_FONT:
                handleQueryFont(reader, writer);
                break;
            case Request.LIST_FONTS:
                handleListFonts(reader, writer);
                break;
            case Request.LIST_FONTS_WITH_INFO:
                handleListFontsWithInfo(client, reader, writer);
                break;
            case Request.QUERY_TEXT_EXTENTS:
                handleQueryTextExtents(reader, writer);
                break;
        }
    }

    private void handleListFontsWithInfo(Client client, PacketReader reader, PacketWriter writer) {
        int max = reader.readCard16();
        int n = reader.readCard16();
        String pattern = reader.readPaddedString(n);
        List<Font> fonts = mFontManager.getFontsMatching(pattern, max);
        final int N = fonts.size();

        for (int i = 0; i < N; i++) {
            PacketWriter pWriter = new PacketWriter(client.getClientListener().getWriter());
            Font font = fonts.get(i);
            String name = font.toString();
            pWriter.setMinorOpCode((byte) name.length());
            writeFont(pWriter, font, N - i);
            pWriter.writePaddedString(name);
            try (PacketReader r = pWriter.copyToReader()) {
                logInfo("FontProtocol", r);
            }
            client.getClientListener().sendPacket(Event.REPLY, pWriter);
        }

        writer.writePadding(52);
    }

    private void handleListFonts(PacketReader reader, PacketWriter writer) {
        int max = reader.readCard16();
        int n = reader.readCard16();
        String pattern = reader.readPaddedString(n);
        List<Font> fonts = mFontManager.getFontsMatching(pattern, max);
        final int N = fonts.size();
        writer.writeCard16(N);
        writer.writePadding(22);
        StringBuilder builder = new StringBuilder();
        for (int i = 0; i < N; i++) {
            String fontName = fonts.get(i).toString();
            builder.append((char) (byte) fontName.length());
            builder.append(fontName);
        }
        writer.writePaddedString(builder.toString());
    }

    private void handleOpenFont(PacketReader reader) {
        int fid = reader.readCard32();
        int length = reader.readCard16();
        reader.readPadding(2);
        String name = reader.readPaddedString(length);
        mFontManager.openFont(fid, name);
    }

    private void handleCloseFont(PacketReader reader) {
        int fid = reader.readCard32();
        mFontManager.closeFont(fid);
    }

    private void handleQueryFont(PacketReader reader, PacketWriter writer) {
        int fid = reader.readCard32();
        Font font = mFontManager.getFont(fid);
        if (font != null) {
            writeFont(writer, font, font.getChars().size());
            writeChars(writer, font);
            try (PacketReader r = writer.copyToReader()) {
                logInfo("FontManager", r);
            }
        } else {
            Log.d("FontManager", "Sending empty font " + Integer.toHexString(fid));
            writer.writePadding(28);
        }
    }

    private void handleQueryTextExtents(PacketReader reader, PacketWriter writer)
            throws GContextError {
        boolean odd = reader.getMinorOpCode() != 0;
        int fid = reader.readCard32();
        Font font = mFontManager.getFont(fid);
        if (font == null) {
            font = mFontManager.getFont(mGraphicsManager.getGc(fid).font);
        }
        int length = (reader.getRemaining() / 2) - (odd ? 1 : 0);
        String s = reader.readString16(length);
        Rect bounds = new Rect();
        int width = (int) font.getPaint().measureText(s);
        font.getPaint().getTextBounds(s, 0, s.length(), bounds);

        writer.writeCard16(font.maxBounds.ascent);
        writer.writeCard16(font.maxBounds.descent);
        writer.writeCard16((short) -bounds.top);
        writer.writeCard16((short) bounds.bottom);
        writer.writeCard32(width);
        writer.writeCard32(bounds.left);
        writer.writeCard32(bounds.right);
        writer.writePadding(4);
    }

    private void writeFont(PacketWriter writer, Font font, int sizeField) {
        writeChar(writer, font.minBounds);
        writer.writePadding(4);
        writeChar(writer, font.maxBounds);
        writer.writePadding(4);

        writer.writeCard16(font.minCharOrByte2);
        writer.writeCard16(font.maxCharOrByte2);
        writer.writeCard16(font.defaultChar);
        final int size = font.mFontProperties.size();
        writer.writeCard16(size);

        writer.writeByte(font.isRtl ? Font.RIGHT_TO_LEFT : Font.LEFT_TO_RIGHT);
        writer.writeByte(font.minByte1);
        writer.writeByte(font.maxByte1);
        writer.writeByte((byte) (font.allCharsExist ? 1 : 0));

        writer.writeCard16(font.fontAscent);
        writer.writeCard16(font.fontDescent);
        writer.writeCard32(sizeField);
        for (int i = 0; i < size; i++) {
            writeProperty(writer, font.mFontProperties.get(i));
        }
    }

    private void writeChars(PacketWriter writer, Font font) {
        List<CharInfo> chars = font.getChars();
        final int N = chars.size();
        for (int i = 0; i < N; i++) {
            writeChar(writer, chars.get(i));
        }
    }

    private void writeProperty(PacketWriter writer, FontProperty property) {
        writer.writeCard32(property.name);
        writer.writeCard32(property.value);
    }

    private void writeChar(PacketWriter writer, Font.CharInfo charInfo) {
        writer.writeCard16(charInfo.leftSideBearing);
        writer.writeCard16(charInfo.rightSideBearing);
        writer.writeCard16(charInfo.characterWidth);
        writer.writeCard16(charInfo.ascent);
        writer.writeCard16(charInfo.descent);
        writer.writeCard16(charInfo.attributes);
    }

    public static void logInfo(String tag, PacketReader r) {
        if (!DEBUG) return;
        Log.d(tag, "minBounds");
        logChar(tag, r);
        Log.d(tag, "");
        Log.d(tag, "maxBounds");
        logChar(tag, r);
        Log.d(tag, "");

        Log.d(tag, "minChar=" + r.readCard16() + " maxChar=" + r.readCard16());
        Log.d(tag, "defaultChar=" + r.readCard16());
        int numProps = r.readCard16();

        Log.d(tag, "");
        Log.d(tag, "isRtl:" + (r.readByte() == Font.RIGHT_TO_LEFT));
        Log.d(tag, "minByte=" + r.readByte() + " maxByte=" + r.readByte());
        Log.d(tag, "allChars:" + (r.readByte() != 0));
        Log.d(tag, "ascent=" + r.readCard16() + " descent=" + r.readCard16());
        Log.d(tag, "");
        Log.d(tag, "Remaining hint: " + r.readCard32());
        for (int i = 0; i < numProps; i++) {
            Log.d(tag, "Prop: " + r.readCard32() + " " + r.readCard32());
        }
        Log.d(tag, "Name: " + r.readPaddedString(r.getMinorOpCode()));
    }

    private static void logChar(String tag, PacketReader r) {
        Log.d(tag, "leftSideBearing=" + r.readCard16()
                + " rightSideBearing=" + r.readCard16());
        Log.d(tag, "width=" + r.readCard16());
        Log.d(tag, "ascent=" + r.readCard16() + " descent=" + r.readCard16());
        Log.d(tag, "attributes=" + r.readCard16());
        r.readPadding(4);
    }
}

File: core/src/main/java/com/monksanctum/xand11/core/fonts/FontProtocol.kt
// Copyright 2018 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package org.monksanctum.xand11.fonts

import org.monksanctum.xand11.comm.*
import org.monksanctum.xand11.core.Font
import org.monksanctum.xand11.core.Platform
import org.monksanctum.xand11.core.Platform.Companion.intToHexString
import org.monksanctum.xand11.core.Rect
import org.monksanctum.xand11.core.Throws
import org.monksanctum.xand11.errors.GContextError
import org.monksanctum.xand11.errors.XError
import org.monksanctum.xand11.core.Font.FontProperty
import org.monksanctum.xand11.fonts.FontManager.Companion.DEBUG
import org.monksanctum.xand11.graphics.GraphicsManager

class FontProtocol(private val mFontManager: FontManager, private val mGraphicsManager: GraphicsManager) : Dispatcher.PacketHandler {

    override val opCodes: ByteArray
        get() = HANDLED_OPS

    @Throws(XError::class)
    override fun handleRequest(client: Client, reader: PacketReader, writer: PacketWriter) {
        when (reader.majorOpCode) {
            Request.OPEN_FONT.code -> handleOpenFont(reader)
            Request.CLOSE_FONT.code -> handleCloseFont(reader)
            Request.QUERY_FONT.code -> handleQueryFont(reader, writer)
            Request.LIST_FONTS.code -> handleListFonts(reader, writer)
            Request.LIST_FONTS_WITH_INFO.code -> handleListFontsWithInfo(client, reader, writer)
            Request.QUERY_TEXT_EXTENTS.code -> handleQueryTextExtents(reader, writer)
            Request.GET_FONT_PATH.code -> handleGetFontPath(reader, writer)
        }
    }

    private fun handleGetFontPath(reader: PacketReader, writer: PacketWriter) {
        writer.writeCard16(0) // 0 Font paths
        writer.writePadding(22)
    }

    private fun handleListFontsWithInfo(client: Client, reader: PacketReader, writer: PacketWriter) {
        val max = reader.readCard16()
        val n = reader.readCard16()
        val pattern = reader.readPaddedString(n)
        val fonts = mFontManager.getFontsMatching(pattern, max)
        val N = fonts.size

        for (i in 0 until N) {
            val pWriter = PacketWriter(client.clientListener.writer)
            val font = fonts[i]
            val name = font.toString()
            pWriter.minorOpCode = name.length.toByte()
            writeFont(pWriter, font, N - i)
            pWriter.writePaddedString(name)
            pWriter.copyToReader().also { r ->
                logInfo("FontProtocol", r)
                r.close()
            }
            client.clientListener.sendPacket(Event.REPLY, pWriter)
        }

        writer.writePadding(52)
    }

    private fun handleListFonts(reader: PacketReader, writer: PacketWriter) {
        val max = reader.readCard16()
        val n = reader.readCard16()
        val pattern = reader.readPaddedString(n)
        val fonts = mFontManager.getFontsMatching(pattern, max)
        val N = fonts.size
        writer.writeCard16(N)
        writer.writePadding(22)
        val builder = StringBuilder()
        for (i in 0 until N) {
            val fontName = fonts[i].toString()
            builder.append(fontName.length.toByte().toChar())
            builder.append(fontName)
        }
        writer.writePaddedString(builder.toString())
    }

    private fun handleOpenFont(reader: PacketReader) {
        val fid = reader.readCard32()
        val length = reader.readCard16()
        reader.readPadding(2)
        val name = reader.readPaddedString(length)
        mFontManager.openFont(fid, name)
    }

    private fun handleCloseFont(reader: PacketReader) {
        val fid = reader.readCard32()
        mFontManager.closeFont(fid)
    }

    private fun handleQueryFont(reader: PacketReader, writer: PacketWriter) {
        val fid = reader.readCard32()
        val font = mFontManager.getFont(fid)
        if (font != null) {
            writeFont(writer, font, font.chars.size)
            writeChars(writer, font)
            writer.copyToReader().also { r ->
                logInfo("FontManager", r)
                r.close()
            }
        } else {
            Platform.logd("FontManager", "Sending empty font " + intToHexString(fid))
            writer.writePadding(28)
        }
    }

    @Throws(GContextError::class)
    private fun handleQueryTextExtents(reader: PacketReader, writer: PacketWriter) {
        val odd = reader.minorOpCode.toInt() != 0
        val fid = reader.readCard32()
        var font: Font? = mFontManager.getFont(fid)
        if (font == null) {
            font = mFontManager.getFont(mGraphicsManager.getGc(fid).font)
        }
        val length = reader.remaining / 2 - if (odd) 1 else 0
        val s = reader.readString16(length)
        val bounds = Rect()
        val width = font!!.measureText(s).toInt()
        font.paintGetTextBounds(s, 0, s.length, bounds)

        writer.writeCard16(font.maxBounds.ascent)
        writer.writeCard16(font.maxBounds.descent)
        writer.writeCard16((-bounds.top).toShort().toInt())
        writer.writeCard16(bounds.bottom.toShort().toInt())
        writer.writeCard32(width)
        writer.writeCard32(bounds.left)
        writer.writeCard32(bounds.right)
        writer.writePadding(4)
    }

    private fun writeFont(writer: PacketWriter, font: Font, sizeField: Int) {
        writeChar(writer, font.minBounds)
        writer.writePadding(4)
        writeChar(writer, font.maxBounds)
        writer.writePadding(4)

        writer.writeCard16(font.minCharOrByte2.toInt())
        writer.writeCard16(font.maxCharOrByte2.toInt())
        writer.writeCard16(font.defaultChar.toInt())
        val size = font.fontProperties.size
        writer.writeCard16(size)

        writer.writeByte(if (font.isRtl) Font.RIGHT_TO_LEFT else Font.LEFT_TO_RIGHT)
        writer.writeByte(font.minByte1)
        writer.writeByte(font.maxByte1)
        writer.writeByte((if (font.allCharsExist) 1 else 0).toByte())

        writer.writeCard16(font.fontAscent)
        writer.writeCard16(font.fontDescent)
        writer.writeCard32(sizeField)
        for (i in 0 until size) {
            writeProperty(writer, font.fontProperties[i])
        }
    }

    private fun writeChars(writer: PacketWriter, font: Font) {
        val chars = font.chars
        val N = chars.size
        for (i in 0 until N) {
            writeChar(writer, chars[i])
        }
    }

    private fun writeProperty(writer: PacketWriter, property: FontProperty) {
        writer.writeCard32(property.name)
        writer.writeCard32(property.value)
    }

    private fun writeChar(writer: PacketWriter, charInfo: Font.CharInfo) {
        writer.writeCard16(charInfo.leftSideBearing)
        writer.writeCard16(charInfo.rightSideBearing)
        writer.writeCard16(charInfo.characterWidth)
        writer.writeCard16(charInfo.ascent)
        writer.writeCard16(charInfo.descent)
        writer.writeCard16(charInfo.attributes)
    }

    companion object {

        private val HANDLED_OPS = byteArrayOf(Request.OPEN_FONT.code,
                Request.QUERY_FONT.code,
                Request.CLOSE_FONT.code,
                Request.LIST_FONTS.code,
                Request.LIST_FONTS_WITH_INFO.code,
                Request.QUERY_TEXT_EXTENTS.code,
                Request.GET_FONT_PATH.code)

        fun logInfo(tag: String, r: PacketReader) {
            if (!DEBUG) return
            Platform.logd(tag, "minBounds")
            logChar(tag, r)
            Platform.logd(tag, "")
            Platform.logd(tag, "maxBounds")
            logChar(tag, r)
            Platform.logd(tag, "")

            Platform.logd(tag, "minChar=" + r.readCard16() + " maxChar=" + r.readCard16())
            Platform.logd(tag, "defaultChar=" + r.readCard16())
            val numProps = r.readCard16()

            Platform.logd(tag, "")
            Platform.logd(tag, "isRtl:" + (r.readByte() == Font.RIGHT_TO_LEFT))
            Platform.logd(tag, "minByte=" + r.readByte() + " maxByte=" + r.readByte())
            Platform.logd(tag, "allChars:" + (r.readByte().toInt() != 0))
            Platform.logd(tag, "ascent=" + r.readCard16() + " descent=" + r.readCard16())
            Platform.logd(tag, "")
            Platform.logd(tag, "Remaining hint: " + r.readCard32())
            for (i in 0 until numProps) {
                Platform.logd(tag, "Prop: " + r.readCard32() + " " + r.readCard32())
            }
            Platform.logd(tag, "Name: " + r.readPaddedString(r.minorOpCode.toInt()))
        }

        private fun logChar(tag: String, r: PacketReader) {
            Platform.logd(tag, "leftSideBearing=" + r.readCard16()
                    + " rightSideBearing=" + r.readCard16())
            Platform.logd(tag, "width=" + r.readCard16())
            Platform.logd(tag, "ascent=" + r.readCard16() + " descent=" + r.readCard16())
            Platform.logd(tag, "attributes=" + r.readCard16())
            r.readPadding(4)
        }
    }
}
